### ISSUE TITLE
Data Preprocessing and Cross-Validation splits.

### ISSUE THREAD

**Issue by Raahul-Singh** [Raahul-Singh/pythia#6]
In the Sunspotter data set, there is a significant bias. The following is a list of the percentage of positive samples per class (not mutually exclusive):

| Forecasting Parameter  | Percentage of Positive samples         |
| ---                         | ---                                        |
|No-flare                 | 24.00%                               |
|C1flr12hr               | 11.37%                               |
|C1flr24hr               | 14.40%                               |
|C5flr12hr               | 8.12%                                 |
|C5flr24hr               | 8.97%                                 |
|M1flr12hr               | 7.22%                                 |
|M1flr24hr               | 7.55%                                 |
|M5flr12hr               | 7.56%                                 |
|M5flr24hr               | 7.66%                                 |

* For dealing with this problem, we can use a method similar to the one described in the paper `Predicting Solar Flares Using a Novel Deep Convolutional Neural Network` [Xuebao Li, Yanfang Zheng, et al 2020](https://iopscience.iop.org/article/10.3847/1538-4357/ab6d04/pdf).
* The gist is:               
    * It is pretty obvious that the number of M-level magnetogram samples is far less than that of No-flare/C-level magnetogram samples, which is consistent with the fact most ARs do not yield major flares in the period of any given 24 hr. This would result in a serious class-imbalance issue, which is a major problem in the field of machine learning.
    * To deal with this, first ARs are categorized into three levels (i.e., No-flare, C, M). “Level = M” indicates that an AR yields at least one M-level flare; “Level = C” indicates that an AR yields at least one C-level flare but no M-level flares “Level = No-flare” indicates that an AR only yields microflares (weaker than C1.0 flares).
    * We shall construct about 10 separate Cross-Validation data sets by the method of shuffle and split Cross-Validation based on AR segregation (AR segregation is, to balance out the classes, I will make sure that the ten Cross-validation splits that I make will have an almost equal number of C, M and No flare producing ARs.). First, we randomly shuffle the AR numbers in different levels of No-flare/C/M and then split the AR numbers at a ratio of around 80%:20% which would correspond to training and testing data respectively.
    * The advantage of this method is that in each of the 10 data sets, not only do the samples in the testing data set not overlap with those in the training data set but also the ARs in the testing data set are disjoint from those in the training data set. We train and evaluate our model on these 10 separate training and testing data sets. We adopt the loss function calculated from the weighted cross-entropy loss.

This can be done by modifying the code from the [repository mentioned in the original paper](https://github.com/FlarePrediction/Repository) to produce the 10 Cross-Validation splits. These modifications would mostly be for making the preprocessing step into a SunPy compliant object.

**Comment by Raahul-Singh** [comment]
Pytorch automatically handles this
